crates/block-producer/src/block_builder/mod.rs

```diff
@@ -197,31 +214,68 @@ impl BlockBuilder {
-    ) -> Result<ProposedBlock, BuildBlockError> {
+    ) -> Result<(ProposedBlock, BlockInputs), BuildBlockError> {
         let BlockBatchesAndInputs { batches, inputs } = batches_inputs;
         let batches = batches.into_iter().map(Arc::unwrap_or_clone).collect();
 
-        let proposed_block =
-            ProposedBlock::new(inputs, batches).map_err(BuildBlockError::ProposeBlockFailed)?;
+        let proposed_block = ProposedBlock::new(inputs.clone(), batches)
+            .map_err(BuildBlockError::ProposeBlockFailed)?;
 
-        Ok(proposed_block)
+        Ok((proposed_block, inputs))
+    }
+
+    #[instrument(target = COMPONENT, name = "block_builder.validate_block", skip_all, err)]
+    async fn validate_block(
+        &self,
+        proposed_block: ProposedBlock,
+        block_inputs: BlockInputs,
+    ) -> Result<(OrderedBatches, BlockInputs, BlockHeader, BlockBody), BuildBlockError> {
+        let response = self
+            .validator
+            .validate_block(proposed_block.clone())
+            .await
+            .map_err(BuildBlockError::ValidateBlockFailed)?;
+
+        // TODO: Check that the returned header and body match the proposed block.
+
+        let (ordered_batches, ..) = proposed_block.into_parts();
+        Ok((ordered_batches, block_inputs, response.header, response.body))
     }
 
     #[instrument(target = COMPONENT, name = "block_builder.prove_block", skip_all, err)]
     async fn prove_block(
         &self,
-        proposed_block: ProposedBlock,
-    ) -> Result<ProvenBlock, BuildBlockError> {
-        let proven_block = self.block_prover.prove(proposed_block).await?;
+        ordered_batches: OrderedBatches,
+        block_inputs: BlockInputs,
+        header: BlockHeader,
+        body: BlockBody,
+    ) -> Result<(OrderedBatches, BlockHeader, BlockBody, BlockProof), BuildBlockError> {
+        // Prove block using header and body from validator.
+        let block_proof = self
+            .block_prover
+            .prove(ordered_batches.clone(), header.clone(), block_inputs)
+            .await?;
+        self.simulate_proving().await;
+
+        Ok((ordered_batches, header, body, block_proof))
+    }
 
+    #[instrument(target = COMPONENT, name = "block_builder.construct_proven_block", skip_all, err)]
+    async fn construct_proven_block(
+        &self,
+        ordered_batches: OrderedBatches,
+        header: BlockHeader,
+        body: BlockBody,
+        block_proof: BlockProof,
+    ) -> Result<ProvenBlock, BuildBlockError> {
+        let proven_block = ProvenBlock::new_unchecked(header, body, block_proof);
         if proven_block.proof_security_level() < MIN_PROOF_SECURITY_LEVEL {
             return Err(BuildBlockError::SecurityLevelTooLow(
                 proven_block.proof_security_level(),
                 MIN_PROOF_SECURITY_LEVEL,
             ));
         }
-
-        self.simulate_proving().await;
+        validate_tx_headers(&proven_block, &ordered_batches.to_transactions())?;
```

Comment:
I'm not sure this is actually needed here - but maybe for now it is fine. Once we have the actual block proving, this would be replaced with verifying the proof returned from the prover against the block header. Maybe worth adding a comment here about this.